Content.Server/MachineLinking/System/SignalLinkerSystem.cs

```diff
@@ -20,262 +20,192 @@ namespace Content.Server.MachineLinking.System
 {
     public sealed class SignalLinkerSystem : EntitySystem
     {
-        private InteractionSystem _interaction = default!;
-
-        private SignalLinkCollection _linkCollection = new();
-
         public override void Initialize()
         {
             base.Initialize();
 
-            _interaction = Get<InteractionSystem>();
-
-            SubscribeLocalEvent<SignalTransmitterComponent, ComponentStartup>(TransmitterStartupHandler);
-            SubscribeLocalEvent<SignalTransmitterComponent, InteractUsingEvent>(TransmitterInteractUsingHandler);
             SubscribeLocalEvent<SignalTransmitterComponent, InvokePortEvent>(OnTransmitterInvokePort);
+            SubscribeLocalEvent<SignalTransmitterComponent, ComponentStartup>(OnTransmitterStartup);
+            SubscribeLocalEvent<SignalTransmitterComponent, ComponentRemove>(OnTransmitterRemoved);
+            SubscribeLocalEvent<SignalTransmitterComponent, InteractUsingEvent>(OnTransmitterInteractUsing);
 
             SubscribeLocalEvent<SignalReceiverComponent, ComponentStartup>(OnReceiverStartup);
+            SubscribeLocalEvent<SignalReceiverComponent, ComponentRemove>(OnReceiverRemoved);
             SubscribeLocalEvent<SignalReceiverComponent, InteractUsingEvent>(OnReceiverInteractUsing);
+        }
 
-            SubscribeLocalEvent<SignalReceiverComponent, ComponentRemove>(OnReceiverRemoved);
-            SubscribeLocalEvent<SignalTransmitterComponent, ComponentRemove>(OnTransmitterRemoved);
+        private void OnTransmitterInvokePort(EntityUid uid, SignalTransmitterComponent component, InvokePortEvent args)
+        {
+            var port = component.Outputs.GetPort(args.Port);
+            foreach (var receiver in port.Receivers)
+                RaiseLocalEvent(receiver.entity, new SignalReceivedEvent(receiver.port), false);
+        }
+
+        private void OnReceiverStartup(EntityUid uid, SignalReceiverComponent component, ComponentStartup args)
+        {
+            if (component.Owner.GetUIOrNull(SignalReceiverUiKey.Key) is { } ui)
+                ui.OnReceiveMessage += msg => OnReceiverUIMessage(uid, component, msg);
+        }
+
```

Comment:
You should just be able to subscribe directly to `SignalPortSelected` events. See how Gas Canister does it.